workflow/controller/operator.go

```diff
@@ -501,20 +503,26 @@ func (woc *wfOperationCtx) persistUpdates() {
 	} else {
 		woc.wf = wf
 	}
+	// After we successfully persist an update to the workflow, the informer's
+	// cache is now invalid. It's very common that we will need to immediately re-operate on a
+	// workflow due to queuing by the pod workers, so we update the informer.
+	// Without this, the next worker to work on this workflow will very likely operate on a stale
+	// object and redo work.
+	un, err := util.ToUnstructured(woc.wf)
+	if err != nil {
+		log.Errorf("failed to convert workflow to unstructured: %v", err)
+	} else {
+		err = woc.controller.wfInformer.GetStore().Update(un)
+		if err != nil {
+			log.Errorf("failed to update workflow: %v", err)
+		}
+	}
 
 	// restore to pre-compressed state
 	woc.wf.Status.Nodes = nodes
 	woc.wf.Status.CompressedNodes = ""
 	woc.log.WithFields(log.Fields{"resourceVersion": woc.wf.ResourceVersion, "phase": woc.wf.Status.Phase}).Info("Workflow update successful")
 
-	// HACK(jessesuen) after we successfully persist an update to the workflow, the informer's
-	// cache is now invalid. It's very common that we will need to immediately re-operate on a
-	// workflow due to queuing by the pod workers. The following sleep gives a *chance* for the
-	// informer's cache to catch up to the version of the workflow we just persisted. Without
-	// this sleep, the next worker to work on this workflow will very likely operate on a stale
-	// object and redo work.
-	time.Sleep(1 * time.Second)
```

Comment:
removing this sleep results in many more updates being processed - we don't necessarily want to suddenly be doing 100s per second

Reply:
We make up for removing this sleep by throttling the the Workflow queue from the Pod informer. Do we also need to account for changes that don't include any Pods? I.e. we could persist a Workflow-only change without any Pods, if it is re-queued soon after are we in danger of operating on a stale workflow?

Reply:
The exchange isn't 1 for 1. This original sleep had the side-effect of batching of updates when many pods changed. The other sleeps goal is to make this implicit side-effect an explicit.